---
title: "Results_Comparison"
output: html_document
date: "`r format(Sys.Date(), '%B %d, %Y')`"
---

```{r install, echo=FALSE, eval=FALSE}
install.packages("VennDiagram")
install.packages("ggvenn")
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
setDTthreads(percent = 90)
library(tidyr)
library(here)
library(ggvenn)
library(DT)
```

## Data comparison

In this script we analyse the intersection of discovered mutations with in both of the analyses

```{r data_readout}
my_data <- fread("all_genes.tsv")
Remy_data <- fread("Remy_Mutations.tsv")
genes_of_interest <- fread(input = "Panel_TSV8.csv")#this is the gene symbol list with all the interesting genes
genes_of_interest <- unique(genes_of_interest$`V8 panel`)
```

And now we'll select only common samples, present in both analyses

```{r common_samples}
common_samples <- unique(intersect(my_data$Sample, Remy_data$Sample_Name))
my_data_common <- my_data[Sample %in% common_samples]
Remy_data_common <- Remy_data[Sample_Name %in% common_samples]
```

Well, now we see `r length(common_samples)` samples present in the both analyses. Here are they `r toString(common_samples)`.  
But of special attention - in the latest analysis we do have `r  nrow(my_data_common)` variants in Exome analysis vs `r nrow(Remy_data_common)` variants in Remy's analysis with chip. Almost 10-times difference!

### Genes abundance

```{r 3-partite Venn diagram, echo=FALSE}
# Example gene symbol sets
genes1 <- unique(my_data_common$symbol)
genes2 <- unique(Remy_data_common$Gene.refGene)


# Prepare the data in a list
gene_sets <- list(
  "Exome analysis" = genes1,
  "Remy analysis" = genes2, 
  "Genes of interest" = genes_of_interest
)

# Create the Venn diagram with custom aesthetics
venn_plot <- ggvenn(
  gene_sets, 
  c("Exome analysis", "Remy analysis", "Genes of interest"),
  fill_color = c("#F8766D", "#00BFC4", "#F0E442"),
  #auto_scale = T, 
  show_percentage = F,
  stroke_size = 0.5,
  set_name_size = 4,
  text_size = 4
)

# Customize the plot with ggplot2
venn_plot <- venn_plot + theme_void() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    plot.background = element_rect(fill = "white")
  ) +
  ggtitle("Venn Diagram of Gene Sets")

print(venn_plot)
```

From this Venn diagram we can see the differences in the gene number in 2 data analyses. "Genes of interest" set - is the gene symbol list from file "Panel_TSV8.csv"\
And we can draw 2 major conclusions here:

-   The exome analysis and previous analysis are not directly comparable, because they comprise different gene numbers

-   To compare the mutations we should remove all the gene symbols not included in the Genes of interest

#### Cleaning-up the genes in the Remy's analysis

```{r Genes clean-up}
Remy_data_goi <- Remy_data_common[Gene.refGene %in% genes_of_interest] 
```

Well, now we see `r nrow(Remy_data_goi)` mutations in the Remy's analysis. It is even fewer than in my analysis - `r  nrow(my_data_common)`! 

And now let's see how it's look like on the Venn diagram  

```{r 2-part Venn diagram, echo=FALSE}

# Prepare the data in a list
gene_sets <- list(
  "Exome analysis" = unique(my_data_common$symbol),
  "Remy analysis" = unique(Remy_data_goi$Gene.refGene) 
)

# Create the Venn diagram with custom aesthetics
venn_plot <- ggvenn(
  gene_sets, 
  c("Exome analysis", "Remy analysis"),
  fill_color = c("#F8766D", "#00BFC4"),
  auto_scale = T, 
  show_percentage = T,
  stroke_size = 0.5,
  set_name_size = 4,
  text_size = 4
)

# Customize the plot with ggplot2
venn_plot <- venn_plot + theme_void() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    plot.background = element_rect(fill = "white")
  ) +
  ggtitle("Venn Diagram of Gene Sets")
print(venn_plot)
```

#### Gene sets comparison

Lets look at the genes, present in the Exome analysis only
```{r Exome_only, echo=FALSE}
common_genes <- unique(intersect(my_data_common$symbol, Remy_data_goi$Gene.refGene))
print(unique(my_data_common$symbol[!(my_data_common$symbol %in% common_genes)]))
```
  
  
Common genes
```{r common_genes, echo=FALSE}
print(common_genes)
```

Well we see here the `r length(common_genes)` - and among them - the most important ones - TP53, KRAS, SMAD4, CDKN2A, FGFR1, BRCA2, at least we are not missing VERY important things! 

  
And genes present only in the Remy's analysis - and NOT detected in exome analysis
```{r chip_only, echo=FALSE}
print(unique(Remy_data_goi$Gene.refGene[!(Remy_data_goi$Gene.refGene %in% common_genes)]))
```

And here we see the gene names we don't have in Exome analysis - in the same cohort of patients. Let's review the list of those genes - their functions and possible implications in cancer. I've made a request in Uniprot.org database and found all of them. So, let's review the results in a table.   

This was the request
```{r, engine='bash', eval=F}
#That is the query submitted to UNIPROT
curl --form 'from="Gene_Name"' \
     --form 'to="UniProtKB"' \
     --form 'ids="MORC4,KDM6A,U2AF1,MLLT4,REV1,ATRNL1,SLAMF7,HSPG2,MUTYH,CD248,BUB1B,MSLN,MEN1,PARP1,FLT3,FGFR2"' \
     --form 'taxId="9606"' \
     https://rest.uniprot.org/idmapping/run
```

What we've obtained - let's have a look

```{r reviewing missed genes}
Remy_only_genes <- fread("Chip_genes_only.tsv")
Remy_only_genes$`Entry Name` <- NULL # Not needed
names(Remy_only_genes) <- c("Gene_symbol", "UNIPROT_ID", "Alt_names", "Protein_length", "Protein_names", "Diseases")
#Cleaning out the things in the "Diseases" column - to make it readable
Remy_only_genes$Diseases <- gsub("\\{.*?\\}", "", Remy_only_genes$Diseases)
setDF(Remy_only_genes)
```
  
  

```{r, Remy_only_genes show, echo=FALSE}
datatable(Remy_only_genes, options = list(pageLength = 20, autoWidth = TRUE), rownames = FALSE)
```

From this table we see 5 genes ASSOCIATED with cancer - MEN1, BUB1B, FLT3, MLLT4, MUTYH - but none of them associated with PDAC. 

#### Variants review

Let's have a closer look at the mutations we potentially missed in the exome analysis and try to find out what has happened
```{r MEN1-MUTYH mutations, echo=FALSE}
datatable(Remy_data_goi[Remy_data_goi$Gene.refGene %in% c("MEN1", "BUB1B", "FLT3", "MLLT4", "MUTYH"), ])
```
  
Well, I can say nothing about these mutations, based on this data. The only thing is it really concerns only these 4 patients `r unique(Remy_data_goi[Remy_data_goi$Gene.refGene %in% c("MEN1", "BUB1B", "FLT3", "MLLT4", "MUTYH"),  Sample_Name])`
  
  
###Variants abundance

Let's review the actual differences in both sets of variations. We'll start with Venn diagrams and then look closer at differences and try to dissect what has happened actually

####Venn diagram
To build the diagrams we actually need to make the variants ID. 
```{r Var_ID_creation}
Remy_data_common$var_ID <- paste("chr", Remy_data_common$Chr, sep = "") #added the prefix chr
Remy_data_common$var_ID <- paste(Remy_data_common$var_ID, Remy_data_common$Start, Remy_data_common$Gene.refGene, sep = "_" )
my_data_common$var_ID <- paste(my_data_common$CHROM, my_data_common$POS, my_data_common$symbol, sep = "_" )
```

Comparison of the set of variants in both analyses, not taking into account the genes of interest

```{r total_variants Venn diagram, echo=FALSE}
# Prepare the data in a list
gene_sets <- list(
  "Exome analysis" = unique(my_data_common$var_ID),
  "Remy analysis" = unique(Remy_data_common$var_ID) 
)

# Create the Venn diagram with custom aesthetics
venn_plot <- ggvenn(
  gene_sets, 
  c("Exome analysis", "Remy analysis"),
  fill_color = c("#F8766D", "#00BFC4"),
  auto_scale = T, 
  show_percentage = T,
  stroke_size = 0.5,
  set_name_size = 4,
  text_size = 4
)

# Customize the plot with ggplot2
venn_plot <- venn_plot + theme_void() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    plot.background = element_rect(fill = "white")
  ) +
  ggtitle("Venn Diagram of Variants intersections")
print(venn_plot)
```
  
  
Well, this is weird. Now I see no point in doing any other comparisons in this setting - because our analyses are clearly incomparable

And the only one variation present in both of the analyses. 
```{r}
intersect(my_data_common$var_ID, Remy_data_common$var_ID)
```

  
  
Nico says that it is the problem in references - that's why the chromosome locations don't match, and as the consequence - the IDs don't match

###Protein level mutations only

In this chunk I'll do the same intersections, but creating IDs in the format like KRAS_GLY12VAL to leverage the references mismatches. 

```{r prot_mut_Remy_table}
#First, with Remy's data -only with genes of interest, we don't bother with others anymore
Remy_goi_mut_table <- Remy_data_goi[, .(Sample_Name, Gene.refGene, GeneDetail.refGene, AAChange.refGene)]
#Parse out the AAchanges fields
Remy_goi_mut_table <- tidyr::separate_longer_delim(Remy_goi_mut_table, cols = "AAChange.refGene", delim = ",")
Remy_goi_mut_table$AAChange.refGene <- sub(".*:p\\.", "", Remy_goi_mut_table$AAChange.refGene)
Remy_goi_mut_table$AAChange.refGene <- sub(".*(wholegene)", "\\1", Remy_goi_mut_table$AAChange.refGene)
Remy_goi_mut_table <- unique(Remy_goi_mut_table)

#Parsing now the field with nucleotide changes
Remy_goi_mut_table <- tidyr::separate_longer_delim(Remy_goi_mut_table, cols = "GeneDetail.refGene", delim = ",")
Remy_goi_mut_table$GeneDetail.refGene <- sub("NM_.*:", "", Remy_goi_mut_table$GeneDetail.refGene)
Remy_goi_mut_table$GeneDetail.refGene <- paste("splicing:", Remy_goi_mut_table$GeneDetail.refGene, sep = "")
setDT(Remy_goi_mut_table)
Remy_goi_mut_table[AAChange.refGene == ".", AAChange.refGene := GeneDetail.refGene]
Remy_goi_mut_table$GeneDetail.refGene <- NULL
Remy_goi_mut_table <- unique(Remy_goi_mut_table)

#Collapsing mutatons into one line
Remy_goi_mut_table[, Remy_Mutations := paste(AAChange.refGene, collapse = "; "), by = .(Gene.refGene, Sample_Name)]
Remy_goi_mut_table <- Remy_goi_mut_table[, .(Sample_Name, Gene.refGene, Remy_Mutations)]
#Renaming columns 
setnames(Remy_goi_mut_table, old = c("Sample_Name", "Gene.refGene"), new = c("Sample", "Gene_symbol"))
Remy_for_merge <- Remy_goi_mut_table[, .(Sample, Gene_symbol, Remy_Mutations)]
```
  
  
Let's have a look what we've got. 
```{r, echo=FALSE}
setDF(Remy_for_merge)
datatable(Remy_for_merge)
```

And let's parse out my table from Exome analysis

```{r prot_mut_Exome_table}
Exome_data_goi_mut_table <- my_data_common[, .(symbol, Sample, Consequence, AA_mutations)]
Exome_data_goi_mut_table <- unique(Exome_data_goi_mut_table)
```

```{r 3-letter to 1-letter conversion}
# This function was made thanks to chatGPT


# Define the conversion function

convert_mutation <- function(mutation) {
  
  # Create a look up table for amino acid codes
  amino_acid_lookup <- c(
  Ala = "A", Arg = "R", Asn = "N", Asp = "D", Cys = "C",
  Gln = "Q", Glu = "E", Gly = "G", His = "H", Ile = "I",
  Leu = "L", Lys = "K", Met = "M", Phe = "F", Pro = "P",
  Ser = "S", Thr = "T", Trp = "W", Tyr = "Y", Val = "V",
  Ter = "*")
  
  # Define patterns for different mutation types
  standard_pattern <- "([A-Za-z]{3})([0-9]+)([A-Za-z]{3})"
  frameshift_pattern <- "([A-Za-z]{3})([0-9]+)([A-Za-z]{3})fsTer([0-9]+)"
  
  # Match against frameshift pattern first
  frameshift_matches <- regmatches(mutation, regexec(frameshift_pattern, mutation))[[1]]
  
  if (length(frameshift_matches) > 0) {
    original_aa <- amino_acid_lookup[frameshift_matches[2]]
    position <- frameshift_matches[3]
    new_aa <- amino_acid_lookup[frameshift_matches[4]]
    fs_position <- frameshift_matches[5]
    converted_mutation <- paste0(original_aa, position, new_aa, "fs*", fs_position)
    return(converted_mutation)
  }
  
  # Match against standard pattern
  standard_matches <- regmatches(mutation, regexec(standard_pattern, mutation))[[1]]
  
  if (length(standard_matches) > 0) {
    original_aa <- amino_acid_lookup[standard_matches[2]]
    position <- standard_matches[3]
    new_aa <- amino_acid_lookup[standard_matches[4]]
    converted_mutation <- paste0(original_aa, position, new_aa)
    return(converted_mutation)
  }
  
  return(NA)
}

Exome_data_goi_mut_table$AA_mut_1_letter <- sapply(Exome_data_goi_mut_table$AA_mutations, convert_mutation)
```

```{r}
Exome_data_goi_mut_table[AA_mutations == ".", AA_mut_1_letter := Consequence]
setnames(Exome_data_goi_mut_table, old = "symbol", new = "Gene_symbol")
Exome_for_merge <- Exome_data_goi_mut_table[, .(Gene_symbol, Sample, AA_mut_1_letter)]
Exome_for_merge <- unique(Exome_for_merge)
Exome_for_merge[, Exome_Mutations := paste(AA_mut_1_letter, collapse = "; "), by = .(Gene_symbol, Sample)]
Exome_for_merge <- unique(Exome_for_merge[, .(Gene_symbol, Sample, Exome_Mutations)])
```

```{r merging_Remy_and_my_results}
comparison_table <- merge(Exome_for_merge, Remy_for_merge)
comparison_table <- unique(comparison_table)
```

Ok, Nelson is happy now. So, for now, I'll export this to excel and separate patient mark it by hands

```{r data export}
fwrite(comparison_table, "Comparison_table.tsv", sep = "\t")
```

####Comparison in numbers

Here we would like to create a table of coincident mutations per patient, like this: 
| Sample | Percentage | total number | Exome_only | Chip_only | common_muts  

```{r mut number comparison}
num_comparison <- comparison_table
num_comparison[, Exome_Mutations := lapply(Exome_Mutations, function(x) strsplit(x, ";")[[1]])]
num_comparison[, Remy_Mutations := lapply(Remy_Mutations, function(x) strsplit(x, ";")[[1]])]
num_comparison[, total_number := length(union(unlist(Exome_Mutations), unlist(Remy_Mutations))), by = Sample]
num_comparison[, common_muts := paste(intersect(unlist(Exome_Mutations), unlist(Remy_Mutations)), collapse = "; "), by = Sample]

#Calculating the percentage of the common genes
num_comparison[, percentage := length(intersect(unlist(Exome_Mutations), unlist(Remy_Mutations))), by = Sample]
num_comparison$percentage <- round(num_comparison$percentage / num_comparison$total_number*100, digits = 1)

num_comparison[, Chip_only := paste(setdiff(unlist(Remy_Mutations), unlist(Exome_Mutations)), collapse = "; "), by = Sample]
num_comparison[, Exome_only := paste(setdiff(unlist(Exome_Mutations), unlist(Remy_Mutations)), collapse = "; "), by = Sample]

#I can add some information to that, but I guess, it is good enough to use

fwrite(num_comparison, "numeric_comparison_ChipvsExome.tsv")
```




























