#Converting funny WES format to table
## Point of this script - to develop a pipeline for converting WES results acrchives to tables
```{r installing_packages, echo=T, results='hide', eval=FALSE}
install.packages("data.table")
install.packages("zip")
install.packages("here")
install.packages("vcfR")
```

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
library(data.table)
setDTthreads(percent = 90)
library(zip)
library(vcfR)
library(here)
```

Here we'll create a function which takes for 
input - the archive names
output - writes files to the desired folder 
Returns nothing
```{r vcf extracting function}
#TODO: add tryCatch - to highlight successful and failed tasks
#This function takes as input archive and outputs the vcf table
#Outputs nothing - just writes files on disk
get_vcf <- function(archive_name, extract_folder = ".", extensions = ".vcf.gz"){
  library(zip)
  file_list <- zip_list(archive_name)$filename
  #Creating a pattern like"(.vcf|.tbi)$"
  pattern <- paste0("(", paste(extensions, collapse = "|"), ")$")
  filename <- file_list[grepl(pattern, file_list, perl = T)]
  unzip(archive_name, exdir = extract_folder, files = filename, junkpaths = T)
  return(invisible(NULL))
}
```

Now let's do that for several files
```{r collecting vcfs, results='hide', message=FALSE}
#collected file names of the archives
PDAC_results <- dir(path="/mnt/sda1/Exomes_June2024",
                    pattern = "^PDAC.*T\\.zip$",
                    full.names = TRUE)

#applying our magical function to these file names
invisible(sapply(PDAC_results, get_vcf, extract_folder = "./exomes_folder"))
```

This is task-specific chunk, we do not need germline variants in this exact study, so we can delete those files
NB! We could be not interested in running it every time, so I've put eval=f
```{r removing genetic variants vcfs, eval=FALSE}
#removing unused vcfs, which we are not interested in
system("rm ./exomes_folder/*_Genetic_Variants.*")
```

```{r renaming_files, engine='bash'}
cd exomes_folder
for i in *.vcf.gz
do
  # Extract the first three consecutive digits found in the filename
  num=$(echo "$i" | grep -oP '\d{3}' | head -n 1)
  
  # Construct the new filename
  TAG="PDAC${num}T.vcf.gz"
  
  # Rename the file
  mv $i $TAG
done
```

This chunk will reheader samples, split vep field and save them as the vcf files 

```{r reheader_split, engine='bash', eval=T}
cd exomes_folder
for i in *.vcf.gz 
do
  TAG=${i%.vcf.gz}
  echo $TAG"_TUMOR" > $TAG"_new_names".txt
  echo $TAG"_NORMAL" >> $TAG"_new_names".txt
  bcftools reheader -s $TAG"_new_names".txt $i | bcftools +split-vep -a ANN -c - -o ${TAG}_VEP_split.vcf
  rm $i $TAG"_new_names".txt
done
echo "Success"
```

This chunk converts vcf files to maf ones and then merges them all to one and compresses it. 
This file will be used later for the oncoplots

Prerequisite - to have vcf2maf.pl installed
NB! This is one-time run script, therefore it is not run - eval=F! 
```{r creating and merging maf files, engine='bash', eval=F}
cd exomes_folder
for i in *.vcf
do
  TAG=${i%_VEP_split.vcf}
  /opt/mskcc-vcf2maf-f6d0c40/vcf2maf.pl --input-vcf $i --inhibit-vep \
  --tumor-id $TAG"_TUMOR" --normal-id $TAG"_NORMAL" \
  --vcf-tumor-id $TAG"_TUMOR" --vcf-normal-id $TAG"_NORMAL" \
  --output-maf ../temp_folder/$TAG.maf --ref-fasta ~/References/GRCh38.p13.genome.fa.gz
done
echo "Success"
```

Merging and archiving maf files
NB! This is one-time run script, therefore it is not run - eval=F! 
```{r merging maf files, engine='bash', eval=F}
cd temp_folder
output_file="concatenated_maf_file.maf"

# Initialize the output file by copying the header from the first MAF file
FIRST_FILE=true

for maf_file in *.maf
do
  if $FIRST_FILE; then
    # Copy the header and content of the first MAF file
    cat "$maf_file" > "$output_file"
    FIRST_FILE=false
  else
    # Skip the header (first line) of subsequent MAF files and append the content
    tail -n +2 "$maf_file" >> "$output_file"
  fi
done

echo "MAF files concatenated into $output_file"

bgzip $output_file
mv $output_file.gz ../exomes_folder
rm *
```


This chunk reads vcfs as tables with vcfR 
This chink runs for about an hour for the ~70 vcf files
```{r merging vcf2tables, eval=FALSE}
vcfs <-  dir(path = "./exomes_folder/", pattern = "*.vcf", full.names = T)

#Producing the table for reference of field names
vcf <- read.vcfR("exomes_folder/PDAC_JIA_HS_003N_VEP_split.vcf")
reference_table <- vcfR2tidy(vcf, single_frame = T)$meta 
rm(vcf)

vcf2table <- function(filename){
tmp <- read.vcfR(filename)
variant_table <- vcfR2tidy(tmp, single_frame = T)$dat
return(variant_table)
}

all_tables <- lapply(vcfs, vcf2table) #the longest step! 
final_table <- rbindlist(all_tables)
rm(all_tables)
invisible(gc())
```

Writing the results to the disk

```{r merging vcf2tables, eval=FALSE}
saveRDS(final_table, file = "exomes_folder/final_table.RDS")
fwrite(final_table, file = "exomes_folder/raw_vcfs.tsv.gz", sep = "\t", compress = "gzip") #saving this on disk, just to be sure
fwrite(reference_table, file = "exomes_folder/field_explanation.tsv", sep = "\t")
file.remove(vcfs) #removed all the vcfs in order not to clog the disk
rm(vcfs, PDAC_results)
```
